sdk/rust/oak/src/http/util.rs

```diff
@@ -0,0 +1,103 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{
+    io::{Receiver, ReceiverExt, Sender, SenderExt},
+    proto::oak::invocation::HttpInvocation,
+    OakError,
+};
+use log::error;
+use oak_abi::label::Label;
+use oak_services::proto::oak::encap::{HttpRequest, HttpResponse};
+
+/// A pair of temporary channels to pass the HTTP request to the Oak Node and receive the response.
+pub struct Pipe {
+    pub request_writer: Sender<HttpRequest>,
+    pub request_reader: Receiver<HttpRequest>,
+    pub response_writer: Sender<HttpResponse>,
+    pub response_reader: Receiver<HttpResponse>,
+}
+
+impl Pipe {
+    pub fn new(request_label: &Label, response_label: &Label) -> Result<Self, OakError> {
+        // Create a channel for passing HTTP requests to the HTTP Client node.
+        let (request_writer, request_reader) =
+            crate::io::channel_create_with_downgrade("HTTP request", request_label)
+                .map_err(OakError::OakStatus)?;
+
+        let (response_writer, response_reader) =
+            crate::io::channel_create_with_downgrade("HTTP response", response_label)
+                .map_err(OakError::OakStatus)?;
+
+        Ok(Pipe {
+            request_writer,
+            request_reader,
+            response_writer,
+            response_reader,
+        })
+    }
+
+    /// Inserts the incoming HTTP request in the `request channel` part of the `HttpInvocation`.
+    pub fn insert_message(&self, request: HttpRequest) -> Result<(), OakError> {
+        // Put the HTTP request message inside the per-invocation request channel.
+        self.request_writer.send_with_downgrade(&request)
+    }
+
+    /// Sends the `HttpInvocation` with request and response channels on the invocation_channel.
+    pub fn send_invocation(
+        &self,
+        invocation_channel: &oak_io::Sender<HttpInvocation>,
+    ) -> Result<(), OakError> {
+        // Create an invocation containing request-specific channels.
+        let invocation = HttpInvocation {
+            receiver: Some(self.request_reader.clone()),
+            sender: Some(self.response_writer.clone()),
+        };
+        invocation_channel.send_with_downgrade(&invocation)
+    }
```

Comment:
The existence of this method and the cloning it performs suggests to me that this Pipe is not the most appropriate abstraction. It seems what we really need is the dual of an Invocation, which has sender and receiver swapped. Then we could create an invocation together with its dual, send the invocation to the other node, and use the dual to read and write to the invocation channels. WDYT?